src/libstd/io/mod.rs

```diff
@@ -1835,6 +1874,51 @@ pub trait BufRead: Read {
         read_until(self, byte, buf)
     }
 
+    /// Read bytes based on a predicate.
+    ///
+    /// `read_while` takes a predicate as an argument.
+    /// It will call this on each byte, and copy it to the slice if the
+    /// predicate evaluates to `true`. Returns the amount of bytes read.
+    ///
+    /// # Errors
+    ///
+    /// If this function encounters an error of the kind
+    /// `ErrorKind::Interrupted` then the error is ignored and the operation
+    /// will continue.
+    ///
+    /// If any other read error is encountered then this function immediately
+    /// returns. Any bytes which have already been read will be appended to
+    /// `buf`.
+    ///
+    /// # Examples
+    ///
+    /// [`std::io::Cursor`][`Cursor`] is a type that implements `BufRead`. In
+    /// this example, we use [`Cursor`] to read bytes in a byte slice until
+    /// we encounter a hyphen:
+    ///
+    /// ```
+    /// use std::io::{self, BufRead};
+    ///
+    /// let mut cursor = io::Cursor::new(b"lorem-ipsum");
+    /// let mut buf = vec![];
+    ///
+    /// cursor.read_while(&mut buf, |b| b != b'-')
+    ///     .expect("reading from cursor won't fail");
+    /// assert_eq!(buf, b"lorem");
+    ///
+    /// let mut buf = vec![];
+    /// cursor.read_to_end(&mut buf);
+    /// assert_eq!(buf, b"-ipsum");
+    /// ```
+    #[unstable(feature = "buf_read_read_while", issue = "none")]
+    fn read_while<P>(&mut self, buf: &mut Vec<u8>, predicate: P) -> Result<usize>
```

Comment:
Probably worth updating read_until to delegate to this method.

Reply:
`read_until` uses `memchr` to scan for the token; it seems likely this operation is vectorized in many cases, and I'm afraid moving to `read_while`'s linear scan could inadvertently cause a slowdown.